### ISSUE TITLE
persist with sync storage

### ISSUE THREAD

**Issue by jbrodriguez** [pmndrs/zustand#478]
Hi,

I'm using react-native-mmkv as a replacement to asyncstorage,

react-native-mmkv is a sync library, but when using it in the persist middlewar, it doesn't rehydrate.

This is what I had to write to get it to work
```
export const storage: StateStorage = {
  getItem: async (key: string): Promise<string | null> => {
    const reply = MMKV.getString(key)
    return Promise.resolve(reply ?? null)
  },
  setItem: async (key: string, value: string): Promise<void> => {
    MMKV.set(key, value)
    Promise.resolve()
  },
}
```

but ideally, it should work synchronously

```
export const storage: StateStorage = {
    getItem: (key: string): string | null => {
      const reply = MMKV.getString(key)
      return reply ?? null
    },
    setItem: (key: string, value: string): void => {
      MMKV.set(key, value)
    },
  }
}

```

I've read a bit about the issue here and I was wondering it https://github.com/pmndrs/zustand/pull/455 fixes the issue, or maybe I should look elsewhere for a fix ?

**Comment by jbrodriguez** [comment]
Update: 3.5.5 solves retrieving the state from onRehydrateStorage, but when I call on function on the returned state, it doesn't work.

```
const useEnvStore = create<Env>(
  persist(
    (set) => ({
      hydrated: false,
      firstRun: true,
      clearFirstRun: () => set({ firstRun: false }),
      setHydrated: () => set({ hydrated: true }),
    }),
    {
      name: 'env',
      getStorage: () => storage,
      whitelist: ['firstRun'],
      onRehydrateStorage: () => (state, error) => {
        if (state) {
          state.setHydrated()
        }
      },
    },
  ),
)
```

In this example, state.setHydrated(), doesn't set hydrated to true.

Any thoughts ?

**Comment by jbrodriguez** [comment]
Ok, worked it out, instead of

`state.setHydrated()` I just do `state.hydrated = true`

it works :)

**Comment by AnatoleLucet** [comment]
It's not a good idea to directly mutate your state's values. `state.setHydrate()` should work. What version of zustand are you using? Can you try with the newest one?

**Comment by jbrodriguez** [comment]
Yes, I have 3.5.5 installed.

With this version, state.setHydrated() inside onRehydrateStorage is not working

**Comment by AnatoleLucet** [comment]
Is `setHydrated` undefined, or is just not setting the state?

**Comment by jbrodriguez** [comment]
It's just not setting the state.

With React Native AsyncStorage it worked normally.

This is how the state looks inside the onRehydrateStorage function

```
state {"clearFirstRun": [Function clearFirstRun], "firstRun": true, "hydrated": false, "setHydrated": [Function setHydrated]}
```

**Comment by AnatoleLucet** [comment]
Strange. Can you make a quick codesandbox so I can reproduce on my end? 🙂

**Comment by jbrodriguez** [comment]
Doing it for react native would be too challenging, expo snack (which is the codesandbox that could be used), doesn't support libraries such as react-native-mmkv yet.

I'll try to set up a quick test using a web sync storage lib

**Comment by AnatoleLucet** [comment]
Oh yeah my bad, it's React Native :confused: 
I've tried to reproduce using your code and`localStorage`, but the issue doesn't seem to appear.
https://codesandbox.io/s/thirsty-goldberg-mb25o?file=/src/App.tsx
![image](https://user-images.githubusercontent.com/35486736/124715356-34e12b80-df03-11eb-8ae5-fe8c89d6e3b6.png)

Do you have an error in `onRehydrateStorage`?

**Comment by jbrodriguez** [comment]
The error param is not set in `onRehydrateStorage` if that's what you meant.

I'll try a clean cache, etc. and retry `setRehydrated`, to check if it was something odd happening.

**Comment by austinried** [comment]
Hi, I have this exact issue after trying to migrate my React Native app from AsyncStorage to mmkv, but I can repro in a codesandbox based off the one posted above:

https://codesandbox.io/s/pedantic-nova-u7fv60?file=/src/App.tsx

I updated it to zustand 3.7.2 and made the storage object sync.  The key difference for me appears to be that I'm using the immer middleware, and I'm guessing that has something to do with it, so I've implemented that in the example as well.

If there's nothing wrong with that example, could this issue be re-opened?

**Comment by austinried** [comment]
Also just to note that immer function I added comes from here: https://github.com/pmndrs/zustand/blob/v3.7.1/tests/middlewareTypes.test.tsx#L20

**Comment by austinried** [comment]
Thanks for that link @AnatoleLucet, I'm not sure how I'd use `hasHydrated` in my case since what I want to do is call a function on the state once hydration has been finished, and that only seems to allow me to check if it's been finished.  I also tried to subscribe to `onFinishHydration` but wasn't able to get that to work, I'm guessing because I was subscribing to it too late and hydration was already done.

But, I did read more into that and learned that sync storage is actually finished when the state is created, so in my case I just switched to calling the function on the state I wanted directly in my splash screen component instead of waiting for hydration and trying to call it in `onRehydrateStorage`/setting my own `hydrated` prop on the state, and now it's working fine.
